File: backend-cf/src/services/ledger.js
/**
 * FlowGuard AI — In-Memory Ledger Service (JS Port)
 */

// ─── Seed Data ────────────────────────────────────────────────────────
const UPI_IDS = [
  'rahul@okaxis', 'priya@ybl', 'amit@paytm', 'sneha@ibl',
  'vikram@kotak', 'deepa@sbi', 'arjun@hdfc', 'kavita@icici',
  'merchant.coffee@okaxis', 'merchant.grocery@ybl',
  'merchant.fuel@paytm', 'merchant.amazon@apl',
  'scammer_suspicious@fake', 'unknown_offshore@intl',
];

const LOCATIONS = [
  'Mumbai', 'Delhi', 'Bangalore', 'Chennai', 'Kolkata',
  'Hyderabad', 'Pune', 'Ahmedabad', 'Dubai', 'Singapore',
  'London', 'New York', 'Lagos', 'Unknown',
];

const DEVICE_TYPES = ['mobile_android', 'mobile_ios', 'web_desktop', 'web_mobile', 'api'];

function uuid() {
  return 'xxxxxxxx'.replace(/x/g, () => Math.floor(Math.random() * 16).toString(16)).toUpperCase();
}

function seededRandom(seed) {
  let s = seed;
  return function () {
    s = (s * 16807) % 2147483647;
    return (s - 1) / 2147483646;
  };
}

function generateSeedTransactions(count = 50) {
  const rand = seededRandom(42);
  const txs = [];
  const now = Date.now();
  const thirtyDays = 30 * 24 * 3600 * 1000;

  for (let i = 0; i < count; i++) {
    const isSuspicious = rand() < 0.15;
    const isCrossBorder = rand() < 0.2;

    let amount, hour, recipient, velocity, location;

    if (isSuspicious) {
      const pick = rand();
      amount = pick < 0.5 ? rand() * 450000 + 50000 : pick < 0.8 ? rand() * 600 + 9900 : rand() * 10 + 1;
      hour = [1, 2, 3, 4, 23, 0][Math.floor(rand() * 6)];
      recipient = rand() < 0.6 ? 'scammer_suspicious@fake' : 'unknown_offshore@intl';
      velocity = Math.floor(rand() * 15) + 5;
      location = ['Unknown', 'Lagos', 'Dubai'][Math.floor(rand() * 3)];
    } else {
      amount = rand() * 24900 + 100;
      hour = Math.floor(rand() * 14) + 8;
      recipient = UPI_IDS[Math.floor(rand() * 12)];
      velocity = Math.floor(rand() * 3);
      location = LOCATIONS[Math.floor(rand() * 8)];
    }

    const txTime = new Date(now - thirtyDays + rand() * thirtyDays);
    txTime.setHours(hour, Math.floor(rand() * 60));
    const currency = !isCrossBorder ? 'INR' : ['USD', 'USDC', 'AED', 'SGD'][Math.floor(rand() * 4)];
    const paymentMethod = !isCrossBorder ? 'upi' : ['stablecoin', 'swift', 'wire'][Math.floor(rand() * 3)];

    txs.push({
      id: `TXN-${uuid()}`,
      amount: Math.round(amount * 100) / 100,
      currency,
      sender: 'user@flowguard',
      recipient,
      payment_method: paymentMethod,
      timestamp: txTime.toISOString(),
      hour,
      location,
      device: DEVICE_TYPES[Math.floor(rand() * DEVICE_TYPES.length)],
      velocity,
      is_cross_border: isCrossBorder,
      is_new_recipient: isSuspicious && rand() > 0.2,
      recipient_freq: isSuspicious ? Math.floor(rand() * 2) : Math.floor(rand() * 8) + 2,
      status: 'completed',
      risk_score: 0,
      risk_level: 'LOW',
      agent_decision: '',
      fee_percentage: Math.round((isCrossBorder ? rand() * 2.9 + 0.1 : rand() * 0.5) * 100) / 100,
    });
  }

  return txs.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
}


export class Ledger {
  constructor() {
    this.users = {
      'user@flowguard': {
        name: 'Aditya Dugar',
        email: 'aditya@flowguard.ai',
        balance: { INR: 250000.00, USDC: 1200.00, USD: 500.00 },
        upi_id: 'aditya@flowguard',
        wallet_address: '0xFlowGuard...Demo',
        risk_profile: 'moderate',
        total_transactions: 156,
        flagged_transactions: 3,
        financial_health_score: 78,
        monthly_spending: 45000,
        monthly_income: 80000,
        savings_rate: 0.44,
      },
    };

    this.transactions = generateSeedTransactions(50);

    const now = new Date();
    this.agentLogs = [
      { timestamp: new Date(now - 300000).toISOString(), action: 'ROUTE_OPTIMIZED', message: 'Routed ₹15,000 via UPI instead of NEFT — saved ₹12 in fees, 23 min faster', confidence: 0.95 },
      { timestamp: new Date(now - 720000).toISOString(), action: 'FRAUD_BLOCKED', message: 'Blocked suspicious ₹4,99,000 to unknown_offshore@intl — unusual amount + location mismatch', confidence: 0.92 },
      { timestamp: new Date(now - 1800000).toISOString(), action: 'INSIGHT_GENERATED', message: 'Your dining expenses are 40% above your monthly average. Consider setting a budget alert.', confidence: 0.88 },
      { timestamp: new Date(now - 3600000).toISOString(), action: 'COMPLIANCE_CHECK', message: 'Transaction to merchant.amazon@apl passed AML/KYC verification — processing normally', confidence: 0.99 },
      { timestamp: new Date(now - 7200000).toISOString(), action: 'STABLECOIN_SWAP', message: 'Converted ₹8,500 → 101.2 USDC for cross-border payment — 60% lower fees than SWIFT', confidence: 0.94 },
    ];

    this.conversionRates = {
      INR_USD: 0.012, USD_INR: 84.15, INR_USDC: 0.012, USDC_INR: 84.10,
      INR_AED: 0.044, AED_INR: 22.90, INR_SGD: 0.016, SGD_INR: 62.50,
      USD_USDC: 1.0, USDC_USD: 1.0,
    };
  }

  getUser(userId = 'user@flowguard') {
    return this.users[userId] || this.users['user@flowguard'];
  }

  getBalance(userId = 'user@flowguard') {
    return this.getUser(userId).balance || {};
  }

  getTransactions(limit = 50, offset = 0) {
    return this.transactions.slice(offset, offset + limit);
  }

  getTransaction(txId) {
    return this.transactions.find(tx => tx.id === txId) || null;
  }

  addTransaction(transaction) {
    if (!transaction.id) transaction.id = `TXN-${uuid()}`;
    if (!transaction.timestamp) transaction.timestamp = new Date().toISOString();
    this.transactions.unshift(transaction);

    const user = this.getUser();
    const currency = transaction.currency || 'INR';
    const amount = transaction.amount || 0;
    if (user.balance[currency] !== undefined) user.balance[currency] -= amount;
    user.total_transactions += 1;

    return transaction;
  }

  addAgentLog(action, message, confidence = 0.9) {
    const log = { timestamp: new Date().toISOString(), action, message, confidence };
    this.agentLogs.unshift(log);
    return log;
  }

  getAgentLogs(limit = 20) {
    return this.agentLogs.slice(0, limit);
  }

  getConversionRate(from, to) {
    return this.conversionRates[`${from}_${to}`] || 1.0;
  }

  getFinancialSummary() {
    const user = this.getUser();
    const totalInr =
      user.balance.INR +
      (user.balance.USDC || 0) * this.conversionRates.USDC_INR +
      (user.balance.USD || 0) * this.conversionRates.USD_INR;

    const flagged = this.transactions.filter(tx => ['HIGH', 'CRITICAL'].includes(tx.risk_level)).length;
    const totalSpent30d = this.transactions
      .filter(tx => tx.currency === 'INR')
      .reduce((sum, tx) => sum + tx.amount, 0);

    return {
      total_balance_inr: Math.round(totalInr * 100) / 100,
      balances: { ...user.balance },
      financial_health_score: user.financial_health_score,
      total_transactions: this.transactions.length,
      flagged_transactions: flagged,
      total_spent_30d: Math.round(totalSpent30d * 100) / 100,
      savings_rate: user.savings_rate,
      risk_profile: user.risk_profile,
    };
  }

  getRiskHeatmapData() {
    const heatmap = {};
    for (const tx of this.transactions) {
      try {
        const dt = new Date(tx.timestamp);
        const key = `${dt.getHours()}_${dt.getDay()}`;
        if (!heatmap[key]) heatmap[key] = { count: 0, total_risk: 0, total_amount: 0, hour: dt.getHours(), day: dt.getDay() };
        heatmap[key].count++;
        heatmap[key].total_risk += tx.risk_score || 0;
        heatmap[key].total_amount += tx.amount || 0;
      } catch { continue; }
    }

    return Object.values(heatmap).map(d => ({
      hour: d.hour,
      day: d.day,
      count: d.count,
      avg_risk: Math.round((d.total_risk / d.count) * 1000) / 1000,
      total_amount: Math.round(d.total_amount * 100) / 100,
    }));
  }
}

File: backend/services/ledger.py
"""
FlowGuard AI — In-Memory Ledger Service
Manages user balances, transaction history, and financial state.
"""
import uuid
from datetime import datetime
from typing import Optional
from data.seed_data import generate_seed_transactions, DEMO_USER, DEMO_AGENT_LOGS


class Ledger:
    """In-memory ledger for payment tracking."""
    
    def __init__(self):
        self.users = {"user@flowguard": {**DEMO_USER}}
        self.transactions = generate_seed_transactions(50)
        self.agent_logs = list(DEMO_AGENT_LOGS)
        self.conversion_rates = {
            "INR_USD": 0.012,
            "USD_INR": 84.15,
            "INR_USDC": 0.012,
            "USDC_INR": 84.10,
            "INR_AED": 0.044,
            "AED_INR": 22.90,
            "INR_SGD": 0.016,
            "SGD_INR": 62.50,
            "USD_USDC": 1.0,
            "USDC_USD": 1.0,
        }
    
    def get_user(self, user_id: str = "user@flowguard") -> dict:
        return self.users.get(user_id, DEMO_USER)
    
    def get_balance(self, user_id: str = "user@flowguard") -> dict:
        user = self.get_user(user_id)
        return user.get("balance", {})
    
    def get_transactions(self, limit: int = 50, offset: int = 0) -> list:
        return self.transactions[offset:offset + limit]
    
    def get_transaction(self, tx_id: str) -> Optional[dict]:
        for tx in self.transactions:
            if tx["id"] == tx_id:
                return tx
        return None
    
    def add_transaction(self, transaction: dict) -> dict:
        """Add a new transaction to the ledger."""
        if "id" not in transaction:
            transaction["id"] = f"TXN-{uuid.uuid4().hex[:8].upper()}"
        if "timestamp" not in transaction:
            transaction["timestamp"] = datetime.now().isoformat()
        
        self.transactions.insert(0, transaction)
        
        # Update balance
        user = self.get_user()
        currency = transaction.get("currency", "INR")
        amount = transaction.get("amount", 0)
        
        if currency in user["balance"]:
            user["balance"][currency] -= amount
        
        user["total_transactions"] += 1
        
        return transaction
    
    def add_agent_log(self, action: str, message: str, confidence: float = 0.9):
        """Add an agent activity log entry."""
        log = {
            "timestamp": datetime.now().isoformat(),
            "action": action,
            "message": message,
            "confidence": confidence,
        }
        self.agent_logs.insert(0, log)
        return log
    
    def get_agent_logs(self, limit: int = 20) -> list:
        return self.agent_logs[:limit]
    
    def get_conversion_rate(self, from_currency: str, to_currency: str) -> float:
        key = f"{from_currency}_{to_currency}"
        return self.conversion_rates.get(key, 1.0)
    
    def get_financial_summary(self) -> dict:
        user = self.get_user()
        total_inr = (
            user["balance"]["INR"] + 
            user["balance"].get("USDC", 0) * self.conversion_rates["USDC_INR"] +
            user["balance"].get("USD", 0) * self.conversion_rates["USD_INR"]
        )
        
        flagged = sum(1 for tx in self.transactions if tx.get("risk_level") in ["HIGH", "CRITICAL"])
        total_spent_30d = sum(tx["amount"] for tx in self.transactions if tx.get("currency") == "INR")
        
        return {
            "total_balance_inr": round(total_inr, 2),
            "balances": user["balance"],
            "financial_health_score": user["financial_health_score"],
            "total_transactions": len(self.transactions),
            "flagged_transactions": flagged,
            "total_spent_30d": round(total_spent_30d, 2),
            "savings_rate": user["savings_rate"],
            "risk_profile": user["risk_profile"],
        }
    
    def get_risk_heatmap_data(self) -> list:
        """Generate risk heatmap data (hour vs day_of_week vs risk)."""
        from collections import defaultdict
        heatmap = defaultdict(lambda: {"count": 0, "total_risk": 0, "total_amount": 0})
        
        for tx in self.transactions:
            try:
                dt = datetime.fromisoformat(tx["timestamp"])
                key = (dt.hour, dt.weekday())
                heatmap[key]["count"] += 1
                heatmap[key]["total_risk"] += tx.get("risk_score", 0)
                heatmap[key]["total_amount"] += tx.get("amount", 0)
            except Exception:
                continue
        
        result = []
        for (hour, day), data in heatmap.items():
            avg_risk = data["total_risk"] / data["count"] if data["count"] > 0 else 0
            result.append({
                "hour": hour,
                "day": day,
                "count": data["count"],
                "avg_risk": round(avg_risk, 3),
                "total_amount": round(data["total_amount"], 2),
            })
        
        return result


# Global instance
ledger = Ledger()
